llpc/lower/llpcSpirvLowerTerminator.cpp

```diff
@@ -124,12 +124,12 @@ void SpirvLowerTerminator::visitCallInst(CallInst &callInst) {
   if (!blockTerm)
     return;
   if (BranchInst *branch = dyn_cast<BranchInst>(blockTerm)) {
-    assert(branch->isUnconditional());
-    BasicBlock *targetBlock = branch->getSuccessor(0);
-    for (PHINode &phiNode : targetBlock->phis()) {
-      for (BasicBlock *incomingBlock : phiNode.blocks()) {
-        if (incomingBlock == parentBlock)
-          return;
+    for (BasicBlock *targetBlock : branch->successors()) {
+      for (PHINode &phiNode : targetBlock->phis()) {
+        for (BasicBlock *incomingBlock : phiNode.blocks()) {
+          if (incomingBlock == parentBlock)
+            return;
+        }
```

Comment:
I realize a lot of this was already there, but there are a number of really strange bits about this code:

1. The check for `blockTerm` -- there has to be a blockTerm in illegal IR, so the check is misleading and can/should be removed. (You could replace it by an assertion if you really want to but personally I don't think that level of paranoia is necessary :))
2. The block terminator could be a `switch` instruction (plus, safest would be to guard against unhandled terminators via `llvm_unreachable`).
3. We unconditionally insert a return instruction with a "null" return value. That's dodgy as hell. How about returning `poison` if the function has a non-void return type?
4. If a successor block contains a phi node, it *must* have `parentBlock` as one of the incoming blocks. Again, if it doesn't, that's invalid IR. So the two inner loops are unnecessary: it's sufficient to check whether the `targetBlock` (better name: `successor`) has a phi node or not.

Also, now that I think about it, even that test isn't quite sufficient because `parentBlock` could be a dominator of its successors, and then no phi nodes are necessary in the first place. Hmm, this is surprisingly tricky to get fully right.

Split the block, insert a conditional branch with constant condition to a block with a `ret` instruction, and let SimplifyCFG take care of it? :)

Reply:
All good points.
I'll have a little bit of a rethink on the whole implementation.

Leaving it all to SimplifyCFG alone cannot necessarily work if the ret containing block relies on values we will skip, etc.
That's said it's OK to just bail on things we cannot handle.
The original purpose of this was to deal with odd stuff like unreachable being added by some compilers to the SPIRV.

Reply:
Reanalyzing the original test case for the branch testing code, it turns out that really what this code was preventing was leaving references to the block containing the kill in PHI nodes.
It seems to be we can always modify this to a return if we updated the PHIs, so I have done this instead.

This now also generates returns of the correct (poison) type.

I tested the SimplifyCFG version as well, and this works for the simple test cases; however, the reason this was written to explicitly erase instruction is to remove any of their references.  If we simply split blocks and leave the work to SimplifyCFG then IR can be invalid leaving this pass as the orphan blocks will references values that they cannot access... Although I guess that is always true of orphan blocks?

Reply:
This is a very nice cleanup. It looks like the "code from the book" for this problem. (I should really start to misappropriate that expression :D)

On the last point: A dominates B if all paths from the entry to B go through A.

If B is reachable, then there are no paths from the entry to B. Therefore, all paths from the entry to B go through A.

Therefore, an unreachable block is dominated by all blocks, and so an unreachable block may refer to values from any other basic block and still be valid SSA form.